aptos-move/aptos-replay-benchmark/src/diff.rs

```diff
@@ -0,0 +1,223 @@
+                    right: None,
+                });
+            }
+        }
+
+        for (right_state_key, right_write_op) in right {
+            diffs.push(Diff::WriteSet {
+                left: None,
+                right: Some((right_state_key, right_write_op)),
+            });
+        }
+    }
+}
+
+impl std::fmt::Display for TransactionDiff {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, " >>>>> ")?;
+        for diff in &self.diffs {
+            match diff {
+                Diff::GasUsed { left, right } => {
+                    writeln!(f, "[gas used] before: {}, after: {}", left, right)?;
+                },
+                Diff::ExecutionStatus { left, right } => {
+                    writeln!(
+                        f,
+                        "[execution status] before: {:?}, after: {:?}",
+                        left, right
+                    )?;
+                },
+                Diff::Event { left, right } => {
+                    let left = left.as_ref();
+                    let right = right.as_ref();
+
+                    if left.is_none() {
+                        writeln!(
+                            f,
+                            "[event] {} was not emitted before",
+                            right.unwrap().type_tag().to_canonical_string()
+                        )?;
+                    } else if right.is_none() {
+                        writeln!(
+                            f,
+                            "[event] {} is not emitted anymore",
+                            left.unwrap().type_tag().to_canonical_string()
+                        )?;
+                    } else {
+                        writeln!(
+                            f,
+                            "[event] {} has changed its data",
+                            left.unwrap().type_tag().to_canonical_string()
+                        )?;
+                    }
+                },
+                Diff::WriteSet { left, right } => {
+                    let left = left.as_ref();
+                    let right = right.as_ref();
+
+                    if left.is_none() {
+                        writeln!(
+                            f,
+                            "[write] {:?} was not written to before",
+                            &right.unwrap().0
+                        )?;
+                    } else if right.is_none() {
+                        writeln!(
+                            f,
+                            "[write] {:?} is not written to anymore",
+                            &left.unwrap().0
+                        )?;
+                    } else {
+                        writeln!(f, "[write] {:?} has changed its value", &left.unwrap().0)?;
+                    }
+                },
+            }
+        }
+        writeln!(f, " <<<<< ")
+    }
+}
+
+// TODO: Add tests here once we allow fine-grained comparisons between events and other parts of
```

Comment:
Can we unit test diff_write_set and diff_events?

Reply:
Done